eval/stringify.ts

```diff
@@ -0,0 +1,176 @@
+ *
+ * When a `Array` or `Object` has a circular reference, a `TypeError` will be thrown.
+ *
+ * ```typescript
+ * import type { Denops } from "jsr:@denops/std";
+ * import { expr, stringify } from "jsr:@denops/std/eval";
+ *
+ * export async function main(denops: Denops): Promise<void> {
+ *   const value = {
+ *     cword: expr`expand('<cword>')`,
+ *     lines: ["foo", "bar"],
+ *   };
+ *   await denops.cmd('echo eval(expr)', { expr: stringify(value) });
+ * }
+ * ```
+ *
+ * @throws {TypeError} If the value contains a circular reference or a `bigint`.
+ */
+export function stringify(value: unknown): string {
+  const ref = new WeakSet();
+  const addRef = (value: unknown[] | Record<PropertyKey, unknown>) => {
+    if (ref.has(value)) {
+      throw new TypeError("Converting circular structure to Vim dict");
+    }
+    ref.add(value);
+  };
+  const reduce = (value: unknown, key: string | number): string => {
+    if (isVimEvaluatable(value)) {
+      return toVimExpression(value);
+    }
+    if (isJsonable(value)) {
+      value = value.toJSON(key);
+      if (isVimEvaluatable(value)) {
+        return toVimExpression(value);
+      }
+    }
+    if (isNullish(value) || isFunction(value) || isSymbol(value)) {
+      return "v:null";
+    }
+    if (isBoolean(value) || isInstanceOfBoolean(value)) {
+      // v:true or v:false
+      return `v:${value}`;
+    }
+    if (isNumber(value) || isInstanceOfNumber(value)) {
+      // Replace `5e-10` to `5.0e-10`
+      return `${value}`.replace(/^(\d+)e/, "$1.0e");
+    }
+    if (isString(value) || isInstanceOfString(value)) {
+      // Vim literal-string
+      return `'${value.replaceAll("'", "''")}'`;
+    }
+    if (value instanceof ArrayBuffer) {
+      // Vim blob
+      return toBlobLiteral(value);
+    }
+    if (isArray(value)) {
+      // Vim list
+      addRef(value);
+      const res = `[${value.map(reduce).join(",")}]`;
+      ref.delete(value);
+      return res;
+    }
+    if (isRecord(value)) {
+      // Vim dict
+      addRef(value);
+      const res = `{${
+        Object.entries(value)
+          .filter(([, value]) => !isIgnoreRecordValue(value))
+          .map(([key, value]) =>
+            `'${key.replaceAll("'", "''")}':${reduce(value, key)}`
+          )
+          .join(",")
+      }}`;
+      ref.delete(value);
+      return res;
+    }
+    const type = Object.prototype.toString.call(value).slice(8, -1);
+    throw new TypeError(`${type} value can't be serialized`);
+  };
+  return reduce(value, "");
+}
+
+type Jsonable = {
+  /**
+   * Returns a JSON value that can be specified to {@linkcode JSON.stringify}.
+   *
+   * @see {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#tojson_behavior|toJSON() behavior}
+   */
+  toJSON(key: string | number): unknown;
+};
+
+function isJsonable(x: unknown): x is Jsonable {
+  return x != null && isFunction((x as Jsonable).toJSON);
+}
```

Comment:
💭 I wonder if we define this in unknownutil.

Reply:
Used [isCustomJsonable in unknownutil@4.3.0](https://jsr.io/@core/unknownutil@4.3.0/doc/~/isCustomJsonable).